protocols/relay/src/v2/protocol/outbound_hop.rs

```diff
@@ -0,0 +1,221 @@
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(HOP_PROTOCOL_NAME)
+    }
+}
+
+impl upgrade::OutboundUpgrade<NegotiatedSubstream> for Upgrade {
+    type Output = Output;
+    type Error = UpgradeError;
+    type Future = BoxFuture<'static, Result<Self::Output, Self::Error>>;
+
+    fn upgrade_outbound(self, substream: NegotiatedSubstream, _: Self::Info) -> Self::Future {
+        let msg = match self {
+            Upgrade::Reserve => HopMessage {
+                r#type: hop_message::Type::Reserve.into(),
+                peer: None,
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+            Upgrade::Connect { dst_peer_id } => HopMessage {
+                r#type: hop_message::Type::Connect.into(),
+                peer: Some(Peer {
+                    id: dst_peer_id.to_bytes(),
+                    addrs: vec![],
+                }),
+                reservation: None,
+                limit: None,
+                status: None,
+            },
+        };
+
+        let mut encoded_msg = Vec::new();
+        msg.encode(&mut encoded_msg)
+            .expect("Vec to have sufficient capacity.");
+
+        let mut codec = UviBytes::default();
+        codec.set_max_len(MAX_MESSAGE_SIZE);
+        let mut substream = Framed::new(substream, codec);
+
+        async move {
+            substream.send(std::io::Cursor::new(encoded_msg)).await?;
+            let msg: bytes::BytesMut = substream
+                .next()
+                .await
+                .ok_or_else(|| std::io::Error::new(std::io::ErrorKind::UnexpectedEof, ""))??;
+
+            let HopMessage {
+                r#type,
+                peer: _,
+                reservation,
+                limit: _,
+                status,
+            } = HopMessage::decode(Cursor::new(msg))?;
+
+            let r#type = hop_message::Type::from_i32(r#type).ok_or(UpgradeError::ParseTypeField)?;
+            match r#type {
+                hop_message::Type::Connect => return Err(UpgradeError::UnexpectedTypeConnect),
+                hop_message::Type::Reserve => return Err(UpgradeError::UnexpectedTypeReserve),
+                hop_message::Type::Status => {}
+            }
+
+            let status = Status::from_i32(status.ok_or(UpgradeError::MissingStatusField)?)
+                .ok_or(UpgradeError::ParseStatusField)?;
+            match status {
+                Status::Ok => {}
+                s => return Err(UpgradeError::UnexpectedStatus(s)),
+            }
+
+            let output = match self {
+                Upgrade::Reserve => {
+                    let reservation = reservation.ok_or(UpgradeError::MissingReservationField)?;
+
+                    let addrs = if reservation.addrs.is_empty() {
+                        return Err(UpgradeError::NoAddressesInReservation);
+                    } else {
```

Comment:
Don't need this else because you are returning above.

Reply:
:+1: removed with 0e1751221c748c8f8d1bb2a0b05f7103e162a6e8.